async-await initial reference material (#635)
rust-lang-nursery/reference: src/expressions/await-expr.md

# Await expressions

> **<sup>Syntax</sup>**\
> _AwaitExpression_ :\
> &nbsp;&nbsp; [_Expression_] `.` `await`

> Await expressions are legal only within `async` contexts, like an

Reviewer: The reference should separately define "async contexts" (also you use backticks inconsistently, either use it everywhere or nowhere) and then refer to it.

`async fn` or an `async` block. They operate on a future. Their effect

> is to "pause" the current computation until the given future is ready

Reviewer: ```suggestion
is to "suspend" the current computation until the given future is ready
```

to produce a value.

> More specifically, a `f.await` expression has the following effect.

Reviewer: ```suggestion
More specifically, an `<expr>.await` expression has the following effect.
```

> 1. evaluate `f` to a future `tmp` using the `IntoFuture:into_future` method;

Reviewer: ```suggestion
1. Evaluate `f` to a future `tmp` using the `IntoFuture::into_future` method;
```

Also there's no such trait...?

Author: hmm yes I guess this changed since the RFC :)

> 1. evaluate `f` to a future `tmp` using the `IntoFuture:into_future` method;

Reviewer 2: ```suggestion
1. evaluate `f` to a future temporary variable `tmp` using the `IntoFuture:into_future` method;
```

I don't like "future" there, but don't really see a better way to succinctly convey the information. It should probably link to the Future trait as well.

> 2. `tmp` is pinned using `Pin::new_unchecked`;

Reviewer: ```suggestion
2. Pin `tmp` using `Pin::new_unchecked`;
```

> 3. This pinned future is then polled by invoking the [`Future::poll`] method and

Reviewer: ```suggestion
3. This pinned future is then polled by calling the [`Future::poll`] method and
```

   passing it the current [task context] (see below);

> 3. if the result returns `Poll::Pending`, then the future returns

Reviewer: ```suggestion
3. If the result evaluates to `Poll::Pending`, then the future returns
```

>    `Poll::Pending`, suspending its state so that, when re-invoked, we will return

Reviewer: ```suggestion
   `Poll::Pending`, suspending its state so that, when re-invoked,
    execution resumes at step 2;
```

Reviewer 2: "re-invoked" to "re-polled"?

Reviewer: Sure, that might be better -- I think both work.

   to step 2;

> 4. if the result returns `Poll::Ready`, then the result is used as the result

Reviewer: ```suggestion
4. If the result evaluates to `Poll::Ready`, then the result is used as the result
```

>    of the await expression itself.

Reviewer: ```suggestion
   of the `await` expression itself.
```

[_Expression_]: expressions.html
[`Future::poll`]: https://doc.rust-lang.org/std/future/trait.Future.html#tymethod.poll
[task context]: https://doc.rust-lang.org/std/task/struct.Context.html

> 

Reviewer: ```suggestion

## Task context

```

The [task context] refers to the context which was supplied to the

> current async block/fn when it itself was polled. Because `await`

Reviewer: ```suggestion
current async context when it itself was polled. Because `await`
```

> current async block/fn when it itself was polled. Because `await`

Reviewer 2: "it itself" should be changed to "the async context" as well. It took me a minute to figure out what "it" was.

expressions are only legal in an async context, there must be some
task context available.

> Effectively, an `f.await` expression desugars into something roughly

Reviewer: ```suggestion
Effectively, an `f.await` expression is roughly
```

# Await expressions

> **<sup>Syntax</sup>**\
> _AwaitExpression_ :\
> &nbsp;&nbsp; [_Expression_] `.` `await`

> Await expressions are legal only within `async` contexts, like an

Reviewer: Can you also link to async context from here?

   of the `await` expression itself.

[`async fn`]: items/functions.html#async-functions
[`async` block]: expressions/block-expr.html#async-blocks
[future]: ../std/future/trait.Future.html
[_Expression_]: expressions.html
[`Future::poll`]: ../std/future/trait.Future.html#tymethod.poll
[`Context`]: ../std/task/struct.Context.html
[`Pin::new_unchecked`]: ../std/pin/struct.Pin.html#method.new_unchecked
[`Poll::Pending`]: ../std/task/enum.Poll.html#variant.Pending
[`Poll::Ready`]: ../std/task/enum.Poll.html#variant.Ready

> **Edition differences**: Await expressions are only available beginning with
> Rust 2018.

## Task context

The task context refers to the [`Context`] which was supplied to the
current [async context] when the async context itself was
polled. Because `await` expressions are only legal in an async
context, there must be some task context available.

[`Context`]: ../std/task/struct.Context.html
[async context]: expressions/block-expr.html#async-context

## Approximate desugaring

Effectively, an `<expr>.await` expression is roughly
equivalent to the following (this desugaring is not normative):

```rust,ignore
loop {

>     let tmp = /* <expr> */;

Reviewer 2: Shouldn't this line be outside of the loop? `tmp` isn't re-evaluated every single time, right? Also, should probably call it `awaited` and not `tmp`.
